src/main/java/com/microsoft/alm/authentication/VsoAzureAuthority.java

```diff
@@ -96,6 +92,56 @@ public VsoAzureAuthority(final String authorityHostUrl)
         return null;
     }
 
+    private URI createPersonalAccessTokenRequestUri(final HttpClient client, final URI targetUri,
+                                                    final boolean requireCompactToken) throws IOException
+    {
+        final String SessionTokenUrl = "_apis/token/sessiontokens?api-version=1.0";
+        final String CompactTokenUrl = SessionTokenUrl + "&tokentype=compact";
+
+        Debug.Assert(client != null, "The client is null");
+
+        final URI identityServiceUri = getIdentityServiceUri(client, targetUri);
+        if (identityServiceUri == null)
+        {
+            throw new RuntimeException("Failed to find Identity Service for " + targetUri.toString());
+        }
+
+        String url = identityServiceUri.toString();
+
+        if (!url.endsWith("/")) {
+            url += "/";
+        }
+
+        url += requireCompactToken ? CompactTokenUrl : SessionTokenUrl;
+
+        return URI.create(url);
+    }
+
+    private URI getIdentityServiceUri(final HttpClient client, final URI targetUri) throws IOException
+    {
+        final String locationServiceUrlFormat = "https://%1$s/_apis/ServiceDefinitions/LocationService2/951917AC-A960-4999-8464-E3F0AA25B381?api-version=1.0";
+
+        Debug.Assert(client != null, ("The client parameter is null."));
+        Debug.Assert(targetUri != null && targetUri.isAbsolute(), "The targetUri parameter is null or invalid");
+
+        final String locationServiceUrl = String.format(locationServiceUrlFormat, targetUri.getHost());
+        URI identityServiceUri = null;
+
+        final HttpURLConnection response = client.get(URI.create(locationServiceUrl));
+        if (response.getResponseCode() == HttpURLConnection.HTTP_OK)
+        {
+            final String responseText = HttpClient.readToString(response);
+
+            identityServiceUri = parseLocationServiceUriFromJson(responseText);
+            if (identityServiceUri != null)
+            {
+                Trace.writeLine("   parsed identity service url: " + identityServiceUri);
+            }

```

Comment:
This naming confused me a little, that `parseLocationServiceUriFromJson()` is used to retrieve the `identityServiceUri`.  Can you add a code comment to explain it?